src/Orleans.CodeGenerator/FieldIdAssignmentHelper.cs

```diff
@@ -0,0 +1,219 @@
+    private readonly LibraryTypes _libraryTypes;
+    private readonly ISymbol[] _memberSymbols;
+    private readonly Dictionary<ISymbol, (ushort, bool)> _symbols = new(SymbolEqualityComparer.Default);
+
+    public bool IsValidForSerialization { get; }
+    public string? FailureReason { get; private set; }
+    public IEnumerable<ISymbol> Members => _memberSymbols;
+
+    public FieldIdAssignmentHelper(INamedTypeSymbol typeSymbol, ImmutableArray<IParameterSymbol> constructorParameters,
+        GenerateFieldIds implicitMemberSelectionStrategy, LibraryTypes libraryTypes)
+    {
+        _constructorParameters = constructorParameters;
+        _implicitMemberSelectionStrategy = implicitMemberSelectionStrategy;
+        _libraryTypes = libraryTypes;
+        _memberSymbols = GetMembers(typeSymbol).ToArray();
+
+        IsValidForSerialization = _implicitMemberSelectionStrategy != GenerateFieldIds.None && !HasMemberWithIdAnnotation() ? GenerateImplicitFieldIds() : ExtractFieldIdAnnotations();
+    }
+
+    public bool TryGetSymbolKey(ISymbol symbol, out (ushort, bool) key) => _symbols.TryGetValue(symbol, out key);
+
+    private bool HasMemberWithIdAnnotation() => _memberSymbols.Any(member => _libraryTypes.IdAttributeTypes.Any(member.HasAttribute));
+
+    private IEnumerable<ISymbol> GetMembers(INamedTypeSymbol symbol)
+    {
+        foreach (var member in symbol.GetMembers().OrderBy(m => m.MetadataName))
+        {
+            if (member.IsStatic || member.IsAbstract)
+            {
+                continue;
+            }
+
+            if (member is not (IFieldSymbol or IPropertySymbol))
+            {
+                continue;
+            }
+
+            if (member.HasAttribute(_libraryTypes.NonSerializedAttribute))
+            {
+                continue;
+            }
+
+            yield return member;
+        }
+    }
+
+    private bool ExtractFieldIdAnnotations()
+    {
+        foreach (var member in _memberSymbols)
+        {
+            if (member is IPropertySymbol prop)
+            {
+                var id = CodeGenerator.GetId(_libraryTypes, prop);
+
+                if (id.HasValue)
+                {
+                    _symbols[member] = (id.Value, false);
+                }
+            }
+
+            if (member is IFieldSymbol field)
+            {
+                var id = CodeGenerator.GetId(_libraryTypes, field);
+                var isConstructorParameter = false;
+
+                if (!id.HasValue)
+                {
+                    var property = GetMatchingProperty(field);
+                    if (property is null)
+                    {
+                        continue;
+                    }
+
+                    id = CodeGenerator.GetId(_libraryTypes, property);
+                    if (!id.HasValue)
+                    {
+                        var constructorParameter = _constructorParameters.FirstOrDefault(x => x.Name.Equals(property.Name, StringComparison.OrdinalIgnoreCase));
```

Comment:
I'm a little confused by this logic. If there's an auto-property then we can imply the id based on the position of a matching constructor parameter, but not if it's a regular field (no matching property). Similarly, the corresponding property member will not have the same id because the logic for `IPropertySymbol` above only checks for an `Id` attribute. Am I missing something?

Reply:
This logic is carried over from the original bits (i.e. how it used to work), but it also confuses me a bit. I think the idea is that fields are more important than properties, so properties are only included if they are explicitly `Id`-annotated, whereas for fields it attempts to first find the corresponding property, and if that fails, attempts to map to a ctor parameter index. The FieldIdAssignmentHelper doesn't know that the code generator prefers fields over properties, it just has logic to attempt to find an id for each member.

Let me know if you want the logic to also check for constructor parameters for properties.. but given how the code generator currently works, I don't think it'd ever get used.